src/webgpu/api/validation/createBindGroupLayout.spec.ts

```diff
@@ -42,185 +41,179 @@ g.test('some binding index was specified more than once', async t => {
-    if (extraTypeSame) {
-      for (const extraBindingType of kBindingTypes) {
-        if (info.perStageLimitType === kBindingTypeInfo[extraBindingType].perStageLimitType) {
-          yield extraBindingType;
-        }
-      }
-    } else {
-      yield info.perStageLimitType === 'sampler' ? 'sampled-texture' : 'sampler';
-    }
-  }
+    t.expectValidationError(() => {
+      t.device.createBindGroupLayout(descriptor);
+    }, !success);
+  });
 
-  kCasesForMaxResourcesPerStageTests = [];
-  for (const maxedType of kBindingTypes) {
-    for (const maxedVisibility of kShaderStages) {
-      // Don't generate a case where maxedType isn't valid in maxedVisibility.
-      if (!(kBindingTypeInfo[maxedType].validStages & maxedVisibility)) continue;
-
-      for (const extraTypeSame of [true, false]) {
-        for (const extraType of pickExtraBindingTypes(maxedType, extraTypeSame)) {
-          for (const extraVisibility of kShaderStages) {
-            // Don't generate a case where extraType isn't valid in extraVisibility.
-            if (!(kBindingTypeInfo[extraType].validStages & extraVisibility)) continue;
-
-            kCasesForMaxResourcesPerStageTests.push({
-              maxedType,
-              maxedVisibility,
-              extraType,
-              extraVisibility,
-            });
-          }
-        }
+// One bind group layout will be filled with kPerStageBindingLimit[...] of the type |type|.
+// For each item in the array returned here, a case will be generated which tests a pipeline
+// layout with one extra bind group layout with one extra binding. That extra binding will have:
+//
+//   - If extraTypeSame, any of the binding types which counts toward the same limit as |type|.
+//     (i.e. 'storage-buffer' <-> 'readonly-storage-buffer').
+//   - Otherwise, an arbitrary other type.
+function* pickExtraBindingTypes(
+  bindingType: GPUBindingType,
+  extraTypeSame: boolean
+): IterableIterator<GPUBindingType> {
+  const info = kBindingTypeInfo[bindingType];
+  if (extraTypeSame) {
+    for (const extraBindingType of kBindingTypes) {
+      if (info.perStageLimitType === kBindingTypeInfo[extraBindingType].perStageLimitType) {
+        yield extraBindingType;
       }
     }
+  } else {
+    yield info.perStageLimitType === 'sampler' ? 'sampled-texture' : 'sampler';
   }
 }
 
+const kCasesForMaxResourcesPerStageTests = params()
+  .combine(poptions('maxedType', kBindingTypes))
+  .combine(poptions('maxedVisibility', kShaderStages))
+  .filter(
+    p => (kBindingTypeInfo[p.maxedType as GPUBindingType].validStages & p.maxedVisibility) !== 0
+  )
+  .expand(function* (p) {
+    for (const extraTypeSame of [true, false]) {
+      yield* poptions('extraType', pickExtraBindingTypes(p.maxedType, extraTypeSame));
+    }
+  })
+  .combine(poptions('extraVisibility', kShaderStages))
+  .filter(
+    p => (kBindingTypeInfo[p.extraType as GPUBindingType].validStages & p.extraVisibility) !== 0
+  );
```

Comment:
notable thing to review

Reply:
Hmm, why is it that `.filter` needs `as` and `expand` doesn't? 

Couldn't `filter(pred: (_: ParamSpec) => boolean)` be `filter(pred: (_: A) => boolean)` ?

Reply:
oops, it shouldn't. I should be able to type-infer that away, yeah.